lib/strzero.c

```diff
@@ -0,0 +1,96 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) Daniel Stenberg, <daniel@haxx.se>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ * SPDX-License-Identifier: curl
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#include "strzero.h"
+
+/*
+ * Curl_explicit_bzero zeroes out a memory buffer in such a way to ensure that
+ * it won't be optimized away by the compiler due to being a dead-store. If no
+ * API for this exists it will fall back to plain memset using a volatile ptr.
+ */
+
+#ifdef HAVE_MEMSET_S
+void
+Curl_explicit_bzero(void *buffer, size_t length)
+{
+  if(!buffer)
+    return;
+  (void)memset_s(buffer, length, 0, length);
+}
+#elif defined(HAVE_BZERO)
+void
+Curl_explicit_bzero(void *buffer, size_t length)
+{
+  if(!buffer)
+    return;
+  bzero(buffer, length);
+}
+#elif defined(HAVE_EXPLICIT_BZERO)
+Curl_explicit_bzero(void *buffer, size_t length)
+{
+  if(!buffer)
+    return;
+  explicit_bzero(buffer, length);
+}
+#elif defined(_WIN32)
+#include <windows.h>
+void
+Curl_explicit_bzero(void *buffer, size_t length)
+{
+  if(!buffer)
+    return;
+  /*
+   * TODO: Windows 11 ships a new function SecureZeroMemory2 which is a safer
```

Comment:
MSDN says the new variant is safe to use in older Windows versions (suggesting this is also an intrinsic):
https://learn.microsoft.com/en-us/windows/win32/memory/winbase-securezeromemory2#remarks
Given the necessary SDK at build-time and doing the detection dance, we might build with it without breaking in older envs.
It requires a recent MSVC; mingw-w64 doesn't support it yet (as of v11).

Reply:
I don't know Windows development even if it jumps up and hits me in the behind (ironic given my employer, I know), so I'm not sure how to use this in a safe way for the various ways to build on Windows.

Reply:
Fair enough! :) I don't have MSVC (or Windows) around, which makes this difficult, so I think the next best thing is to wait till someone figures it out. For now, it's so new that GitHub shows zero actual use.